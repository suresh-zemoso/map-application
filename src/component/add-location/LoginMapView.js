import React, { useState } from 'react';
import BaseMapView from '../public-map/BaseMapView';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import { useSelector } from "react-redux";
import { locationService } from '../../services/locationService';





const useStyles = makeStyles({
    sidebarStyle: {
        position: 'absolute',
        backgroundColor: '#404040',
        color: '#ffffff',
        zIndex: '1!important',
        padding: '6px',
        fontWeight: 'bold'
    }
});

export default function LoginMapView(props) {

    const [mapAttribute, setMapAttribute] = useState({ lng: 73.413827, lat: 25.676973, zoom: 2 });
    const [open, setOpen] = React.useState(false);

    const accessToken = useSelector(state => state.authentication.accessToken);
    const classes = useStyles(props);

    const handleClickOpen = () => {
        setOpen(true);
    };

    const handleClose = () => {
        setOpen(false);
    };

    function addMapOnclick(map) {
        map.on('click', (e) => {
            setMapAttribute({
                lng: e.lngLat.lng.toFixed(6),
                lat: e.lngLat.lat.toFixed(6),
                zoom: map.getZoom().toFixed(2)
            });
        });
    }
    function addLocation(mapAttribute) {
        const data = {
            lng: mapAttribute.lng,
            lat: mapAttribute.lat
        }
        locationService.addLocation(data, accessToken)
            .then(response => {

            })
            .catch(error => {
                console.log(error);
            })
        handleClose();
    }
    return (
        <>
            <BaseMapView
                handlers={[addMapOnclick]}
                mapAttribute={{
                    lng: mapAttribute.lng,
                    lat: mapAttribute.lat,
                    zoom: mapAttribute.zoom
                }}
                mapMode="streets-v11"
                buttons={[{ buttonText: "Add Location", onClick: () => handleClickOpen() }]}
            >
                <div className={classes.sidebarStyle}>
                    <div>Longitude: {mapAttribute.lng} | Latitude: {mapAttribute.lat}</div>
                </div>
            </BaseMapView >
            <Dialog
                open={open}
                onClose={handleClose}
                aria-labelledby="alert-dialog-title"
                aria-describedby="alert-dialog-description"
            >
                <DialogTitle id="alert-dialog-title">{"Save Location"}</DialogTitle>
                <DialogContent>
                    <DialogContentText id="alert-dialog-description">
                        Do you want to save location with Longitude {mapAttribute.lng} and Latitude {mapAttribute.lat} ?
          </DialogContentText>
                </DialogContent>
                <DialogActions>
                    <Button onClick={handleClose} color="primary">
                        Cancel
          </Button>
                    <Button onClick={() => addLocation(mapAttribute)} color="primary" autoFocus>
                        Confirm
          </Button>
                </DialogActions>
            </Dialog>
        </>
    )
}
