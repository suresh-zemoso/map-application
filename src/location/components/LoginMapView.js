import React, { useState, useRef } from 'react';
import BaseMapView from '../../location/components/BaseMapView';
import { makeStyles } from '@material-ui/core/styles';
import { useSelector, useDispatch } from "react-redux";
import Snackbar from '@material-ui/core/Snackbar';
import Alert from '@material-ui/lab/Alert';
import { addLocation } from '../../location/actions/locationAction';
import mapboxgl from 'mapbox-gl';
import PropTypes from 'prop-types';




const useStyles = makeStyles({
    sidebarStyle: {
        position: 'absolute',
        backgroundColor: '#404040',
        color: '#ffffff',
        zIndex: '1!important',
        padding: '6px',
        fontWeight: 'bold'
    },
    marker: {
        background: `url(${require("./mapbox-icon.svg")})`,
        backgroundRepeat: 'no-repeat',
        backgroundSize: '45px 45px',
        width: '50px',
        height: '50px',
        borderRadius: '50%',
        cursor: 'pointer'
    },
    snackbar: {
        left: '63% !important'
    }
});


const defaultProps = {

}

const propTypes = {


}
const LoginMapView = (props) => {

    const [mapAttribute, setMapAttribute] = useState({ lng: 73.413827, lat: 25.676973, zoom: 2 });
    const [isSnackbarOpen, setSnackbarOpen] = useState(false);
    const [messageInfo, setMessageInfo] = useState(undefined)
    const queueRef = useRef([])

    const accessToken = useSelector(state => state.authentication.accessToken);
    const saveError = useSelector(state => state.locationState.saveError);

    const dispatch = useDispatch();

    const classes = useStyles(props);

    function formatMessage(lat, lng) {
        return `Location with Latitude ${lat} and Longitude ${lng} Added`
    }

    function processQueue() {
        if (queueRef.current.length > 0) {
            setMessageInfo(queueRef.current.shift());
            setSnackbarOpen(true);
        }
    };

    function openSnackbar(message) {
        queueRef.current.push({
            message,
            key: new Date().getTime(),
        });

        if (isSnackbarOpen) {
            setSnackbarOpen(false);
        } else {
            processQueue();
        }
    };

    function handleClose(event, reason) {
        if (reason === 'clickaway') {
            return;
        }
        setSnackbarOpen(false);
    };

    function handleExited() {
        processQueue();
    };

    function mousemoveListener(map, e) {
        setMapAttribute({
            lng: e.lngLat.lng.toFixed(6),
            lat: e.lngLat.lat.toFixed(6),
            zoom: map.getZoom().toFixed(2)
        });
    }

    function onClickListener(map, e) {
        // create a HTML element for each feature
        var el = document.createElement('div');
        el.className = classes.marker;
        console.debug("Lat lang value", e.lngLat.lat, e.lngLat.lng)
        // make a marker for each feature and add to the map
        const location = {
            lng: e.lngLat.lng.toFixed(6),
            lat: e.lngLat.lat.toFixed(6)
        }

        dispatch(addLocation(location, accessToken))
            .then(data => {
                openSnackbar(formatMessage(location.lat, location.lng));
                new mapboxgl.Marker(el)
                    .setLngLat([e.lngLat.lng, e.lngLat.lat])
                    .addTo(map);
            },
                error => {
                    openSnackbar("Error in adding location");
                    console.log(error)
                })
    }

    const handlers = [{
        addEventListener: (map) => map.on('mousemove', (e) => mousemoveListener(map, e)),
        removeEventListener: (map) => map.off('mousemove', (e) => mousemoveListener(map, e))
    },
    {
        addEventListener: (map) => map.on('click', (e) => onClickListener(map, e)),
        removeEventListener: (map) => map.off('click', (e) => onClickListener(map, e))
    }
    ]


    return (
        <div>
            <BaseMapView
                handlers={handlers}
                mapAttribute={{
                    lng: mapAttribute.lng,
                    lat: mapAttribute.lat,
                    zoom: mapAttribute.zoom
                }}
                mapMode="streets-v11"
            >
                <div className={classes.sidebarStyle}>
                    <div>Longitude: {mapAttribute.lng} | Latitude: {mapAttribute.lat}</div>
                </div>
            </BaseMapView >
            <Snackbar
                key={messageInfo ? messageInfo.key : undefined}
                anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
                open={isSnackbarOpen}
                onExited={handleExited}
                classes={{ anchorOriginTopCenter: classes.snackbar }}
                // message={messageInfo ? messageInfo.message : undefined}
                // autoHideDuration={3000}
                onClose={handleClose} >
                <Alert severity={saveError ? "error" : "success"}>
                    {messageInfo ? messageInfo.message : undefined}
                </Alert>
            </Snackbar>
        </div>
    )
}

LoginMapView.propTypes = propTypes;
LoginMapView.defaultProps = defaultProps;

export default LoginMapView;