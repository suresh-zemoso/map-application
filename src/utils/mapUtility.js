function getGeoCoder(mapboxgl) {
    return new window.MapboxGeocoder({
        accessToken: mapboxgl.accessToken,
        mapboxgl: mapboxgl
    })
}

function getGeoLocateControl(mapboxgl) {
    return new mapboxgl.GeolocateControl({
        positionOptions: {
            enableHighAccuracy: true
        },
        trackUserLocation: true
    })
}

export const mapUtility = { getGeoCoder, getGeoLocateControl }