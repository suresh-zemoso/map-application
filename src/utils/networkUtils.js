import axios from 'axios';

export function handleApiResponse(response) {


    const data = response && response.data;

    if (response.status !== 200 && response.status !== 201) {
        if (response.status === 401) {
            console.debug("Unauthorized got excecuted")
        }
        const error = (data && data.message) || response;
        return Promise.reject(error);
    }

    return data;

}

axios.interceptors.response.use((response) => {
    return response;
}, function (error) {
    // Do something with response error
    return Promise.reject(error.response);
});