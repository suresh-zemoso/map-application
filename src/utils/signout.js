import { logout } from "../auth/actions/loginAction"
import { store } from "../store";

export default function signout(props) {
    store.dispatch(logout())
}
