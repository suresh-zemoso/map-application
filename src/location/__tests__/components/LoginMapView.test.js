import React from 'react';
import { mount, shallow } from 'enzyme';
import LoginMapView from '../../components/LoginMapView';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as ReactReduxHooks from '../../../hooks/reduxHooks';
import { Provider } from 'react-redux';
import { mapUtility } from '../../../utils/mapUtility';


jest.mock('mapbox-gl', () => {
    const mockMap = jest.fn(() => ({
        on: jest.fn(),
        off: jest.fn(),
        getSource: jest.fn(),
        getLayer: jest.fn(),
        addControl: jest.fn(),
        addSource: jest.fn(),
        addLayer: jest.fn(),
        addImage: jest.fn(),
        loadImage: jest.fn(),
        hasImage: jest.fn(),
        dragPan: {
            enable: jest.fn(),
            disable: jest.fn()
        },
        scrollZoom: {
            enable: jest.fn(),
            disable: jest.fn()
        },
        doubleClickZoom: {
            enable: jest.fn(),
            disable: jest.fn()
        },
        touchZoomRotate: {
            enable: jest.fn(),
            disable: jest.fn()
        },
        setStyle: jest.fn(),
        queryRenderedFeatures: jest.fn(),
        getCenter: jest.fn(),
        flyTo: jest.fn(),
        isStyleLoaded: jest.fn(),
        getBounds: jest.fn(),
        getZoom: jest.fn(),
        resize: jest.fn(),
        getStyle: jest.fn(() => ({
            layers: []
        })),
        isMoving: jest.fn(),
        isZooming: jest.fn(),
        isRotating: jest.fn(),
        removeControl: jest.fn(),
        remove: jest.fn()
    }));

    const mockPopup = jest.fn(() => ({
        setLngLat: jest.fn(() => mockPopup()),
        setDOMContent: jest.fn(() => mockPopup()),
        remove: jest.fn(() => mockPopup()),
        addTo: jest.fn(() => mockPopup()),
        setMaxWidth: jest.fn(() => mockPopup())
    }));

    const mockMarker = jest.fn(() => ({
        setLngLat: jest.fn(() => mockMarker()),
        remove: jest.fn(() => mockMarker()),
        addTo: jest.fn(() => mockMarker()),
        getElement: jest.fn(() => mockMarker()),
        on: jest.fn(),
        getLngLat: jest.fn()
    }));

    return {
        Map: mockMap,
        NavigationControl: jest.fn(),
        GeolocateControl: jest.fn(() => ({
            on: jest.fn(),
            trigger: jest.fn()
        })),
        Popup: mockPopup,
        LngLat: jest.fn((lat, lng) => ({
            lat, lng,
            wrap: jest.fn(() => ({
                lat, lng
            }))
        })),
        Marker: mockMarker,
        ScaleControl: jest.fn()
    };
});


describe('LoginMapView', () => {

    let wrapper;
    let store;

    beforeEach(() => {

        /* mocking store */
        store = configureStore([thunk])({
            authentication: {},
            locationState: {}
        });

        jest
            .spyOn(ReactReduxHooks, "useSelector")
            .mockImplementation(state => store.getState());

        jest.spyOn(mapUtility, "getGeoCoder").mockImplementation(() => { })

        /* shallow rendering */
        wrapper = mount(<Provider store={store}><LoginMapView store={store} /></Provider>);
    });


    it('should render the LoginMapView Component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

});      