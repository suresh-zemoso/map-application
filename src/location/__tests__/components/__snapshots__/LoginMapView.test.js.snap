// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginMapView should render the LoginMapView Component correctly 1`] = `ReactWrapper {}`;
